File: WalletKitCore/src/walletkit/WKListener.c
//
//  WKListener.c
//  WK
//
//  Created by Ed Gamble on 8/11/20
//  Copyright © 2020 Breadwinner AG. All rights reserved.
//
//  See the LICENSE file at the project root for license information.
//  See the CONTRIBUTORS file at the project root for a list of contributors.

#include "WKListenerP.h"
#include "support/BROSCompat.h"

#include "WKNetwork.h"
#include "WKTransfer.h"
#include "WKWallet.h"
#include "WKWalletManager.h"
#include "WKSystem.h"

IMPLEMENT_WK_GIVE_TAKE (WKListener, wkListener)

// MARK: - Generate Transfer Event

typedef struct {
    BREvent base;
    WKListener listener;
    WKWalletManager manager;
    WKWallet wallet;
    WKTransfer transfer;
    WKTransferEvent event;
} BRListenerSignalTransferEvent;

static void
wkListenerSignalTransferEventDispatcher (BREventHandler ignore,
                                             BRListenerSignalTransferEvent *event) {
    event->listener->transferCallback (event->listener->context,
                                       event->manager,
                                       event->wallet,
                                       event->transfer,
                                       event->event);
}

static BREventType handleListenerSignalTransferEventType = {
    "CWM: Handle Listener Transfer EVent",
    sizeof (BRListenerSignalTransferEvent),
    (BREventDispatcher) wkListenerSignalTransferEventDispatcher
};

private_extern void
wkListenerGenerateTransferEvent (const WKTransferListener *listener,
                                     WKTransfer transfer,
                                     WKTransferEvent event) {
    if (NULL == listener || NULL == listener->listener) return;

    BRListenerSignalTransferEvent listenerEvent =
    { { NULL, &handleListenerSignalTransferEventType},
        listener->listener,
        wkWalletManagerTakeWeak (listener->manager),
        wkWalletTakeWeak (listener->wallet),
        wkTransferTakeWeak (transfer),
        event };

    eventHandlerSignalEvent(listener->listener->handler, (BREvent *) &listenerEvent);
}

// MARK: - Generate Wallet Event

typedef struct {
    BREvent base;
    WKListener listener;
    WKWalletManager manager;
    WKWallet wallet;
    WKWalletEvent event;
} BRListenerSignalWalletEvent;

static void
wkListenerSignalWalletEventDispatcher (BREventHandler ignore,
                                           BRListenerSignalWalletEvent *event) {
    event->listener->walletCallback (event->listener->context,
                                     event->manager,
                                     event->wallet,
                                     event->event);
}

static BREventType handleListenerSignalWalletEventType = {
    "CWM: Handle Listener Wallet Event",
    sizeof (BRListenerSignalWalletEvent),
    (BREventDispatcher) wkListenerSignalWalletEventDispatcher
};

private_extern void
wkListenerGenerateWalletEvent (const WKWalletListener *listener,
                                   WKWallet wallet,
                                   OwnershipGiven WKWalletEvent event) {
    if (NULL == listener || NULL == listener->listener) return;

    BRListenerSignalWalletEvent listenerEvent =
    { { NULL, &handleListenerSignalWalletEventType},
        listener->listener,
        wkWalletManagerTakeWeak (listener->manager),
        wkWalletTakeWeak (wallet),
        event };

    eventHandlerSignalEvent(listener->listener->handler, (BREvent *) &listenerEvent);
}

// MARK: - Generate Manager Event

typedef struct {
    BREvent base;
    WKListener listener;
    WKWalletManager manager;
    WKWalletManagerEvent event;
} BRListenerSignalManagerEvent;

static void
wkListenerSignalManagerEventDispatcher (BREventHandler ignore,
                                            BRListenerSignalManagerEvent *event) {
    event->listener->managerCallback (event->listener->context,
                                      event->manager,
                                      event->event);
}

static BREventType handleListenerSignalManagerEventType = {
    "CWM: Handle Listener Manager Event",
    sizeof (BRListenerSignalManagerEvent),
    (BREventDispatcher) wkListenerSignalManagerEventDispatcher
};

private_extern void
wkListenerGenerateManagerEvent (const WKWalletManagerListener *listener,
                                    WKWalletManager manager,
                                    WKWalletManagerEvent event) {
    if (NULL == listener || NULL == listener->listener) return;

    BRListenerSignalManagerEvent listenerEvent =
    { { NULL, &handleListenerSignalManagerEventType},
        listener->listener,
        wkWalletManagerTakeWeak (manager),
        event };

    eventHandlerSignalEvent (listener->listener->handler, (BREvent *) &listenerEvent);
}

// MARK: - Generate Network Event

typedef struct {
    BREvent base;
    WKListener listener;
    WKNetwork network;
    WKNetworkEvent event;
} BRListenerSignalNetworkEvent;

static void
wkListenerSignalNetworkEventDispatcher (BREventHandler ignore,
                                            BRListenerSignalNetworkEvent *event) {
    event->listener->networkCallback (event->listener->context,
                                      event->network,
                                      event->event);
}

static BREventType handleListenerSignalNetworkEventType = {
    "CWM: Handle Listener Network Event",
    sizeof (BRListenerSignalNetworkEvent),
    (BREventDispatcher) wkListenerSignalNetworkEventDispatcher
};

private_extern void
wkListenerGenerateNetworkEvent (const WKNetworkListener *listener,
                                    WKNetwork network,
                                    WKNetworkEvent event) {
    if (NULL == listener || NULL == listener->listener) return;

    BRListenerSignalNetworkEvent listenerEvent =
    { { NULL, &handleListenerSignalNetworkEventType},
        listener->listener,
        wkNetworkTakeWeak (network),
        event };

    eventHandlerSignalEvent (listener->listener->handler, (BREvent *) &listenerEvent);
}

// MARK: - Generate System Event

typedef struct {
    BREvent base;
    WKListener listener;
    WKSystem system;
    WKSystemEvent event;
} BRListenerSignalSystemEvent;

static void
wkListenerSignalSystemEventDispatcher (BREventHandler ignore,
                                           BRListenerSignalSystemEvent *event) {
    event->listener->systemCallback (event->listener->context,
                                     event->system,
                                     event->event);
}

static BREventType handleListenerSignalSystemEventType = {
    "CWM: Handle Listener System Event",
    sizeof (BRListenerSignalSystemEvent),
    (BREventDispatcher) wkListenerSignalSystemEventDispatcher
};

private_extern void
wkListenerGenerateSystemEvent (WKListener listener,
                                   WKSystem system,
                                   WKSystemEvent event) {
    if (NULL == listener) return;

    BRListenerSignalSystemEvent listenerEvent =
    { { NULL, &handleListenerSignalSystemEventType},
        listener,
        wkSystemTakeWeak (system),
        event };

    eventHandlerSignalEvent (listener->handler, (BREvent *) &listenerEvent);
}

// MARK: - Event Type

static const BREventType *
wkListenerEventTypes[] = {
    &handleListenerSignalNetworkEventType,
    &handleListenerSignalTransferEventType,
    &handleListenerSignalWalletEventType,
    &handleListenerSignalManagerEventType,
    &handleListenerSignalSystemEventType
};

static const unsigned int
wkListenerEventTypesCount = (sizeof (wkListenerEventTypes) / sizeof(BREventType*)); //  11

extern WKListener
wkListenerCreate (WKListenerContext context,
                      WKListenerSystemCallback systemCallback,
                      WKListenerNetworkCallback networkCallback,
                      WKListenerWalletManagerCallback managerCallback,
                      WKListenerWalletCallback walletCallback,
                      WKListenerTransferCallback transferCallback) {
    
    WKListener listener = calloc (1, sizeof (struct WKListenerRecord));

    listener->ref = WK_REF_ASSIGN (wkListenerRelease);
    pthread_mutex_init_brd (&listener->lock, PTHREAD_MUTEX_NORMAL_BRD);

    listener->context          = context;
    listener->systemCallback   = systemCallback;
    listener->networkCallback  = networkCallback;
    listener->managerCallback  = managerCallback;
    listener->walletCallback   = walletCallback;
    listener->transferCallback = transferCallback;

    listener->handler = eventHandlerCreate ("Core SYS, Listener",
                                            wkListenerEventTypes,
                                            wkListenerEventTypesCount,
                                            &listener->lock);

    return listener;
}

static void
wkListenerRelease (WKListener listener) {
    eventHandlerStop (listener->handler);
    eventHandlerDestroy (listener->handler);

    pthread_mutex_destroy (&listener->lock);

    memset (listener, 0, sizeof(*listener));
    free (listener);
}

extern void
wkListenerStart (WKListener listener) {
    eventHandlerStart (listener->handler);
}

extern void
wkListenerStop (WKListener listener) {
    eventHandlerStop (listener->handler);
}

File: WalletKitJava/WalletKitNative/src/main/java/com/blockset/walletkit/nativex/WKListener.java
/*
 * Created by Michael Carrara <michael.carrara@breadwallet.com> on 7/1/19.
 * Copyright (c) 2019 Breadwinner AG.  All right reserved.
 *
 * See the LICENSE file at the project root for license information.
 * See the CONTRIBUTORS file at the project root for a list of contributors.
 */
package com.blockset.walletkit.nativex;

import com.blockset.walletkit.nativex.library.WKNativeLibraryDirect;
import com.blockset.walletkit.nativex.utility.Cookie;
import com.sun.jna.Callback;
import com.sun.jna.Pointer;
import com.sun.jna.PointerType;

public class WKListener extends PointerType {

    //
    // Implementation Detail
    //
    public interface BRCryptoListenerSystemEvent extends Callback {
        void callback(Pointer context,
                      Pointer system,
                      WKSystemEvent.ByValue event);
    }

    public interface BRCryptoListenerNetworkEvent extends Callback {
        void callback(Pointer context,
                      Pointer network,
                      WKNetworkEvent.ByValue event);
    }

    public interface BRCryptoListenerWalletManagerEvent extends Callback {
        void callback(Pointer context,
                      Pointer manager,
                      WKWalletManagerEvent.ByValue event);
    }

    public interface BRCryptoListenerWalletEvent extends Callback {
        void callback(Pointer context,
                      Pointer manager,
                      Pointer wallet,
                      WKWalletEvent event);
    }

    public interface BRCryptoListenerTransferEvent extends Callback {
        void callback(Pointer context,
                      Pointer manager,
                      Pointer wallet,
                      Pointer transfer,
                      WKTransferEvent.ByValue event);
    }

    //
    // Client Interface
    //

    public interface SystemEventCallback extends BRCryptoListenerSystemEvent {
        void handle(Cookie context,
                /* OwnershipGiven */ WKSystem system,
                /* OwnershipGiven */ WKSystemEvent.ByValue event);

        @Override
        default void callback(Pointer context,
                              Pointer system,
                              WKSystemEvent.ByValue event) {
            handle(new Cookie(context),
                    new WKSystem(system),
                    event);
        }
    }

    public interface NetworkEventCallback extends BRCryptoListenerNetworkEvent {
        void handle(Cookie context,
                /* OwnershipGiven */ WKNetwork network,
                /* OwnershipGiven */ WKNetworkEvent.ByValue event);

        @Override
        default void callback(Pointer context,
                              Pointer network,
                              WKNetworkEvent.ByValue event) {
            handle(new Cookie(context),
                    new WKNetwork(network),
                    event);
        }
    }

    public interface WalletManagerEventCallback extends BRCryptoListenerWalletManagerEvent {
        void handle(Cookie context,
                /* OwnershipGiven */ WKWalletManager manager,
                /* OwnershipGiven */ WKWalletManagerEvent.ByValue event);

        @Override
        default void callback(Pointer context,
                              Pointer manager,
                              WKWalletManagerEvent.ByValue event) {
            handle(new Cookie(context),
                    new WKWalletManager(manager),
                    event);
        }
    }

    public interface WalletEventCallback extends BRCryptoListenerWalletEvent {
        void handle(Cookie context,
                /* OwnershipGiven */ WKWalletManager manager,
                /* OwnershipGiven */ WKWallet wallet,
                /* OwnershipGiven */ WKWalletEvent event);

        @Override
        default void callback(Pointer context,
                              Pointer manager,
                              Pointer wallet,
                              WKWalletEvent event) {
            handle(new Cookie(context),
                    new WKWalletManager(manager),
                    new WKWallet(wallet),
                    event);
        }
    }

    public interface TransferEventCallback extends BRCryptoListenerTransferEvent {
        void handle(Cookie context,
                /* OwnershipGiven */ WKWalletManager manager,
                /* OwnershipGiven */ WKWallet wallet,
                /* OwnershipGiven */ WKTransfer transfer,
                /* OwnershipGiven */ WKTransferEvent.ByValue event);

        @Override
        default void callback(Pointer context,
                              Pointer manager,
                              Pointer wallet,
                              Pointer transfer,
                              WKTransferEvent.ByValue event) {
            handle(new Cookie(context),
                    new WKWalletManager(manager),
                    new WKWallet(wallet),
                    new WKTransfer(transfer),
                    event);
        }
    }

    //
    // Listener Pointer
    //

    public WKListener() {
        super();
    }

    public WKListener(Pointer pointer) {
        super(pointer);
    }

    public static WKListener create(Cookie context,
                                    SystemEventCallback systemEventCallback,
                                    NetworkEventCallback networkEventCallback,
                                    WalletManagerEventCallback walletManagerEventCallback,
                                    WalletEventCallback walletEventCallback,
                                    TransferEventCallback transferEventCallback) {
        return new WKListener(
                WKNativeLibraryDirect.wkListenerCreate(
                        context.getPointer(),
                        systemEventCallback,
                        networkEventCallback,
                        walletManagerEventCallback,
                        walletEventCallback,
                        transferEventCallback));
    }

    public WKListener take() {
        Pointer thisPtr = this.getPointer();

        return new WKListener(WKNativeLibraryDirect.wkListenerTake(thisPtr));
    }

    public void give() {
        Pointer thisPtr = this.getPointer();

        WKNativeLibraryDirect.wkListenerGive(thisPtr);
    }

}
